src/core/search/compressed_sorted_set.cc

```diff
@@ -0,0 +1,223 @@
+  ptrdiff_t diff_offset = bound.diff_span.data() - diffs_.data();
+  size_t required_len = diff1_span.size() + diff2_span.size();
+  DCHECK_LE(bound.diff_span.size(), required_len);  // It can't shrink for sure
+  diffs_.insert(diffs_.begin() + diff_offset, required_len - bound.diff_span.size(), 0u);
+
+  // Now overwrite diff0 and 0s with the two new differences
+  copy(diff1_span.begin(), diff1_span.end(), diffs_.begin() + diff_offset);
+  copy(diff2_span.begin(), diff2_span.end(), diffs_.begin() + diff_offset + diff1_span.size());
+}
+
+// Remove has linear complexity. It tries to find the element V and its neighbors A and B,
+// which are encoded as diff1 = V - A and diff2 = B - V. Adjacently stored diff1 and diff2
+// need to be replaced with diff3 = diff1 + diff2s
+void CompressedSortedSet::Remove(IntType value) {
+  auto bound = LowerBound(value);
+
+  // Nothing was read or the element was not found
+  if (bound.diff_span.empty() || bound.value != value)
+    return;
+
+  // We're removing below unconditionally
+  size_--;
+
+  // Calculate offset where values diff is stored and determine diffs tail
+  ptrdiff_t diff_offset = bound.diff_span.data() - diffs_.data();
+  auto diffs_tail = absl::MakeSpan(diffs_).subspan(diff_offset + bound.diff_span.size());
+
+  // If it's stored at the end, simply truncate it away
+  if (diffs_tail.empty()) {
+    diffs_.resize(diffs_.size() - bound.diff_span.size());
+    return;
+  }
+
+  // Now the list certainly contains a succeeding element B > V and possibly A < V (or 0)
+  // Read diff2 and calculate diff3 = diff1 + diff2
+  auto [diff2, diff2_read] = ReadVarLen(diffs_tail);
+  IntType diff3 = (bound.value - bound.prev_value) + diff2;
+
+  // Encode diff3
+  VarintBuffer buf;
+  auto diff3_buf = WriteVarLen(diff3, absl::MakeSpan(buf));
+
+  // Shrink vector before overwriting
+  DCHECK_LE(diff3_buf.size(), diff2_read + bound.diff_span.size());
+  size_t to_remove = diff2_read + bound.diff_span.size() - diff3_buf.size();
+  diffs_.erase(diffs_.begin() + diff_offset, diffs_.begin() + diff_offset + to_remove);
+
+  // Overwrite diff1/diff2 with new diff3
+  copy(diff3_buf.begin(), diff3_buf.end(), diffs_.begin() + diff_offset);
+}
+
+size_t CompressedSortedSet::Size() const {
+  return size_;
+}
+
+size_t CompressedSortedSet::ByteSize() const {
+  return diffs_.size();
+}
+
+// The leftmost three bits of the first byte store the number of additional bytes. All following
+// bits store the number itself.
+absl::Span<uint8_t> CompressedSortedSet::WriteVarLen(IntType value, absl::Span<uint8_t> buf) {
+  buf[0] = value & 0b1'11'11;
+  value >>= 5;
+
+  uint8_t tail_size = 0;
+  while (value > 0) {
+    buf[++tail_size] = value & 0xFF;
+    value >>= 8;
+  }
+
+  DCHECK_LE(tail_size, 4);
+  buf[0] |= (tail_size << 5);
+  return buf.first(tail_size + 1);
+}
+
+std::pair<CompressedSortedSet::IntType, size_t> CompressedSortedSet::ReadVarLen(
+    absl::Span<const uint8_t> source) {
```

Comment:
I suggest using flit encoding from helio (see flit.h).
1. It's more efficient in space  for small integers (requires 1 byte for 0-127)
2. It's more efficient with CPU - it does not have loops and a very small number of branches.

see `flit::Encode32` and `flit::Parse64Safe`

Reply:
in this case you can even use `flit::Parse64Fast` - there is no risk of the vector not being 8-byte aligned.

Reply:
haha, this is exactly the encoding I proposed earlier :)
1 question though: when you say "8-byte aligned" - you don't mean memory alignment, right? Because the vector has integers (potentially) starting with all 8-byte alignment options. I guess you meant that there's always going to be 8 more bytes to read, but that's not necessarily true when reading the last part of the array, right?

Reply:
Wow, didn't know we have it. Great, its even not used anywhere in DF so far

Reply:
@chakaz  it's not google varint encoding, it's something else. what have you proposed?

I actually meant that with an allocated byte array, it will always be 8-byte aligned so if for example, we have only a single byte in the array, the process can still access 8 bytes starting from that pointer (yes, 7 others are garbage) without risking to segfault. Parse64Fast has a potential risk of segfaulting because it always loads 8 bytes before parsing.

Reply:
@romange I realize this is not Google varint, I proposed an encoding very similar to this (but bits are slightly flipped), see https://github.com/dragonflydb/dragonfly/pull/1474#discussion_r1241122264

> it will always be 8-byte aligned
Oh, I thought that you were talking about memory alignment (like data structures), sorry

> we have only a single byte in the array, the process can still access 8 bytes starting from that pointer (yes, 7 others are garbage)
What fulfills this guarantee? The data is allocated in a `vector` on the heap